JSON now used for statefile backends; datreant.data moved to its own package

dotsdl opened the pull request:
This PR contains three major changes to the library. Namely,

1. JSON is now the only backend for state files used in the library. See issue #31 for the advantages/disadvantages of this move. Closes issue #31.
2. `datreant.data` is no longer included in the base package, but has been moved to [its own repo](https://github.com/datreant/datreant.data).
3. `Treant._attach_limb()` is now used for dynamically attaching limbs, such as `Treant.tags`, as properties. This mechanism is also used by [`datreant.data`](https://github.com/datreant/datreant.data) for attaching `Treant.data` and `Bundle.data` limbs.

These changes mean that the base `datreant` package now requires no HDF5-related dependencies, as well as simplifying the state file model it's built upon. It also sets up the pattern for future development, in which specialized but perhaps dependency-heavy functionality can be built independently without bloating the core package.

Other *minor* changes:

1. `Coordinator` object has been removed. This object was never really developed, and may make less sense going forward. It has not been removed from mention among the docs, though. There were a lot of unused + untested code elements that were built to facilitate this object, and these were removed.
2. Switched from absolute to relative imports whenever possible.
3. `six` is now a dependency, and we use `string_types` instead of `str` whenever we do an `isinstance` check.
4. `numpy` and `pandas` are no longer dependencies, and we no longer need to imitate the structure of PyTables tables using structured arrays in objects like `Bundle`.
5. Added some initial backend tests for state files in `test_backends.py`
6. Added builds for python 3.3, 3.4, 3.5 to travis CI instructions.

A [script](https://github.com/datreant/pytables-to-json/blob/master/pytables-to-json.py) for creating JSON state files from the now-deprecated HDF5 state files is ready for use.

kain88-de commented:
Do you mind splitting this into 3 PR's one for each change? I didn't have time yet to really dig into the library so having 3 big things in one is a little bit hard to review.

dotsdl commented:
I'll try. This one kind of snowballed and took on a mind of its own. I do have [another branch](https://github.com/datreant/datreant/tree/exp-json_only) in which I removed all the old backends in favor of only a JSON one, off of which this branch is based, so I'll set that up as a PR. That probably has a lot of bugs that I ended up fixing along the way in this one, however.

dotsdl commented:
I've made a new PR that includes only the change from PyTables to JSON: #34

kain88-de commented on datreant/__init__.py:
```diff
@@ -12,15 +12,12 @@
-import datreant.data
+from .treants import Treant, Group, register
+from .collections import Bundle
+from .manipulators import *
```
grab all imports are discouraged by PEP8.

dotsdl replied:
Agreed. Removing `manipulators` as an import for now. There's only the one (`datreant.manipulators.get()`), and we should explicitly include these anyhow.

dotsdl replied:
Done.

kain88-de commented on datreant/backends/core.py:
```diff
@@ -316,3 +313,115 @@ def _close(self):
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock:
```
shouldn't that check if the `fdlock` is set to 'shared'? Can a lock for example be exclusive? Could it be in the future?

dotsdl replied:
This is a check for when **this** process already has any kind of lock on the file, either 'shared' or 'exclusive' (which happens for writes). For reading, this process can have either kind of lock already before it does any kind of reading. This check allows a method that itself already has a lock on the file to call another method that would first get a lock. We don't want the called method to change the lock, so this just lets it proceed.

kain88-de replied:
could you rename this decorator to `_read_and_lock`. That would make it clearer in the code where the decorator is applied that the file is locked as well. It's not much but could make code using the decorator easier to understand for new devs.

dotsdl replied:
See below. Going to combine the decorators to make things less confusing.

dotsdl replied:
Done. Since these decorators are always used in a particular way, there was no use having four of them. There are now only `_read` and `_write`.

kain88-de commented on datreant/backends/core.py:
```diff
@@ -316,3 +313,115 @@ def _close(self):
+                try:
+                    out = func(self, *args, **kwargs)
+                finally:
+                    self._unlock(self.fd)
```
this looks like it might be good to thing to put into a function called `release_lock` you use this 3 lines several times in the same manner.

dotsdl replied:
Good point. I think these three lines are never used in another order elsewhere, but I'll check. If that's the case it makes perfect sense to move them into their own method.

dotsdl replied:
Done.

kain88-de commented:
So I had a rough look over it. My comments are mostly about stuff where I noticed there could be room for improvement. I can't say much about the actual implementation. Unfortunately I didn't look through the code much yet.

dotsdl commented:
Thanks a ton for doing it! So far this package has had mostly just my eyes scrolling over it, and as you know, you don't see the problems that you're blind to. All of this is making the package better. :D

kain88-de commented on datreant/backends/core.py:
```diff
@@ -316,3 +329,139 @@ def _close(self):
+        return inner
+
+    @staticmethod
+    def _write(func):
```
same thing goes for this function

dotsdl replied:
Done!

kain88-de commented on datreant/backends/core.py:
```diff
@@ -316,3 +329,139 @@ def _close(self):
+        that changes to the data are written to the file after the method
+        returns.
+
+        This decorator doesn't do any file locking, but should be used after
```
Ok this means that the order of decorators might be important. So for example only one of the following two functions uses the right declaration. I'm actually not sure which one it is.

```python
@_read_and_lock
@_pull_push
def foo():
    pass


@_pull_push
@_read_and_lock
def bar():
     pass
```

So I get this is a hard thing to avoid the way you do this right now. But I see that this might be an easy error to make then class methods are declared. A easy thing to do is change the name to something like `_pull_push_needs_lock`, this should work as a reminder for people who review the code that the decorator ordering is important.

Unfortunately I don't know of a way to automatically check the order of applied decorators.

Of course ideally the order doesn't matter but we can't do that here.

dotsdl replied:
I think the better thing to do here is to combine `_read` with `_pull` and `_write` with `_pull_push`, since these operations in practice always come together in a specific order. I agree that as written right now it's more than a little confusing.

kain88-de replied:
Why dont you just rename them like I suggested and decorate the pull push functions within?

kain88-de replied:
I like it better when you decorate the push pull functions. It keeps them nice and short plus the locking is stated very obvious when one looks at the definition of the function.

dotsdl replied:
I went ahead and combined them. I think that keeps things simpler, since they really are never used apart.

kain88-de commented on datreant/backends/statefiles.py:
```diff
@@ -0,0 +1,448 @@
+            else:
+                raise
+
+    def _init_record(self):
```
from where is this function called? Somewhere in the base class?

kain88-de replied:
if not then it might be better to move this code to the `__init__` function as well.

dotsdl replied:
It's used to initialize the data structure in case it doesn't exist. When a `TreantFile` object is created, it first creates its proxy file for applying locks, and then calls its `create` method, which will then call methods like `update_schema`, `update_version`, `add_tags`, and `add_categories`. However, these all require a write to the state file, but what if the file doesn't actually exist (such as when this is a new `Treant`)? Pulling the current version of the data structure will fail when each of these methods is called.

A smooth way to make this work is to have the `_write` decorator handle it. A lock gets applied to the proxy file, but if the statefile itself doesn't exist yet, the data structure it would have is initialized, and this is later written out when the method that was wrapped finishes.

See [here](https://github.com/datreant/datreant/blob/exp-dynamic_data/datreant/backends/core.py#L401) for where it lives in the `_write` decorator. I think there are other ways of handling the init issue for the data structure that lives in the state file, but I think this one has nice features.

kain88-de commented on datreant/backends/statefiles.py:
```diff
@@ -0,0 +1,448 @@
+    def _init_record(self):
+        self._record = dict()
+        self._record['tags'] = list()
+        self._record['categories'] = dict()
```
and to the creation with a dict

```python
self._record = {'tags': list(), 'categories': dict()}
```

it saves 2 lines of code and expresses the same.

dotsdl replied:
Done!

kain88-de commented on datreant/backends/statefiles.py:
```diff
@@ -0,0 +1,448 @@
+        # add new tags
+        self._record['tags'].extend(tags)
+
+    @FileSerial._write
```
yeah if the first decorator would state in it's name that a lock is acquired would be nice.

dotsdl replied:
I think the naming is a lot clearer now. This kind of grew out of the needs of the old PyTables backend, so the cleanup was certainly necessary.

kain88-de commented on datreant/backends/statefiles.py:
```diff
@@ -0,0 +1,448 @@
+
+        """
+        # check if uuid already present
+        uuids = [member[0] for member in self._record['members']]
```
didn't you say this is a namedtuple now? or is gthub just behind in displaying the diff?

dotsdl replied:
That's the other PR #34 (yeah, never doing overlapping PRs ever again; sorry about this). :/

Each member record is now a dict instead of a list, so this problem will go away. Considering merging the changes made for that PR into this branch, but afraid that might just make things more confusing for the moment.

kain88-de commented on datreant/backends/statefiles.py:
```diff
@@ -0,0 +1,448 @@
+                dict giving full member data, with fields as keys and in member
+                order
+        """
+        out = {key: [] for key in self._fields}
```
defaultdict

dotsdl replied:
Done!

kain88-de commented on datreant/collections.py:
```diff
@@ -152,6 +167,48 @@ def names(self):
+                list giving the basedir of each member, in order;
+                members that are missing will have basedir ``None``
+
+        """
```
what should also work is

```python
basedirs = [member.basedir if member else None for member in self._list()]
```

This comprehension did something similar in a short test
```python
[i if i %2 ==0 else None for i in range(8)]
```

dotsdl replied:
Done!

kain88-de commented on datreant/collections.py:
```diff
@@ -321,17 +347,13 @@ def add_member(self, uuid, treanttype, basedir):
+        uuids = [member[0] for member in self.record]
+
+        if uuid not in uuids:
+            self.record.append([uuid,
```
namedtuple member

dotsdl replied:
See above. Other branch that this one builds on now uses dicts for member records instead of lists. This still works since the member record is a list of dicts, maintaining ordering.

kain88-de commented on datreant/collections.py:
```diff
@@ -1,23 +1,23 @@
 import glob
 import fnmatch
 
+from six import string_types
```
also do a `from six.moves import zip` because you use zip later in this file. It will make sure you are using izip in python2 and 3.

dotsdl replied:
Done!

kain88-de commented:
No problem. Finding problems in other peoples code is easy ;-). 

Sorry for the mess with the comments in the commits before. But that seemed to me to be the easiest way to go through your changes and have some idea what has been done.

dotsdl commented:
Addressed all the comments. I think after PR #34 is merged into `develop` I'll merge into this branch the changes made there. We can then check that I didn't miss anything in fixing merge conflicts. Altogether it's looking a lot better!

dotsdl commented:
Alright, I merged the changes made in `develop` from PR #34 into this branch, so now the only major change here is the removal of `datreant.data` and the introduction of `Treant._attach_limb` and `Bundle._attach_limb` for adding limbs as new properties to these objects.

We should wait for the verdict of #35 before merging this, since without a namespace package we can't expect imports of [`datreant.data`](https://github.com/datreant/datreant.data) to work as a separate package.
